// parse-server/lib/GraphQL/loaders/functionsMutations.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.load = void 0;

var _graphql = require("graphql");

var _FunctionsRouter = require("../../Routers/FunctionsRouter");

var defaultGraphQLTypes = _interopRequireWildcard(require("./defaultGraphQLTypes"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

const load = parseGraphQLSchema => {
  const fields = {};
  fields.call = {
    description: 'The call mutation can be used to invoke a cloud code function.',
    args: {
      functionName: {
        description: 'This is the name of the function to be called.',
        type: new _graphql.GraphQLNonNull(_graphql.GraphQLString)
      },
      params: {
        description: 'These are the params to be passed to the function.',
        type: defaultGraphQLTypes.OBJECT
      }
    },
    type: defaultGraphQLTypes.ANY,

    async resolve(_source, args, context) {
      try {
        const {
          functionName,
          params
        } = args;
        const {
          config,
          auth,
          info
        } = context;
        return (await _FunctionsRouter.FunctionsRouter.handleCloudFunction({
          params: {
            functionName
          },
          config,
          auth,
          info,
          body: params
        })).response.result;
      } catch (e) {
        parseGraphQLSchema.handleError(e);
      }
    }

  };
  const functionsMutation = new _graphql.GraphQLObjectType({
    name: 'FunctionsMutation',
    description: 'FunctionsMutation is the top level type for functions mutations.',
    fields
  });
  parseGraphQLSchema.graphQLTypes.push(functionsMutation);
  parseGraphQLSchema.graphQLMutations.functions = {
    description: 'This is the top level for functions mutations.',
    type: functionsMutation,
    resolve: () => new Object()
  };
};

exports.load = load;
