// parse-server/lib/GraphQL/loaders/usersQueries.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.load = void 0;

var _graphql = require("graphql");

var _graphqlListFields = _interopRequireDefault(require("graphql-list-fields"));

var _node = _interopRequireDefault(require("parse/node"));

var _rest = _interopRequireDefault(require("../../rest"));

var _Auth = _interopRequireDefault(require("../../Auth"));

var _parseClassTypes = require("./parseClassTypes");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const load = parseGraphQLSchema => {
  if (parseGraphQLSchema.isUsersClassDisabled) {
    return;
  }

  const fields = {};
  fields.me = {
    description: 'The Me query can be used to return the current user data.',
    type: new _graphql.GraphQLNonNull(parseGraphQLSchema.meType),

    async resolve(_source, _args, context, queryInfo) {
      try {
        const {
          config,
          info
        } = context;

        if (!info || !info.sessionToken) {
          throw new _node.default.Error(_node.default.Error.INVALID_SESSION_TOKEN, 'Invalid session token');
        }

        const sessionToken = info.sessionToken;
        const selectedFields = (0, _graphqlListFields.default)(queryInfo);
        const {
          include
        } = (0, _parseClassTypes.extractKeysAndInclude)(selectedFields);
        const response = await _rest.default.find(config, _Auth.default.master(config), '_Session', {
          sessionToken
        }, {
          include: include.split(',').map(included => `user.${included}`).join(',')
        }, info.clientVersion);

        if (!response.results || response.results.length == 0 || !response.results[0].user) {
          throw new _node.default.Error(_node.default.Error.INVALID_SESSION_TOKEN, 'Invalid session token');
        } else {
          const user = response.results[0].user;
          user.sessionToken = sessionToken;
          return user;
        }
      } catch (e) {
        parseGraphQLSchema.handleError(e);
      }
    }

  };
  const usersQuery = new _graphql.GraphQLObjectType({
    name: 'UsersQuery',
    description: 'UsersQuery is the top level type for users queries.',
    fields
  });
  parseGraphQLSchema.graphQLTypes.push(usersQuery);
  parseGraphQLSchema.graphQLQueries.users = {
    description: 'This is the top level for users queries.',
    type: usersQuery,
    resolve: () => new Object()
  };
};

exports.load = load;
